ray-operator/controllers/ray/rayservice_controller.go

```diff
@@ -72,3 +158,203 @@ func (r *RayServiceReconciler) SetupWithManager(mgr ctrl.Manager) error {
+		ObjectMeta: metav1.ObjectMeta{
+			Labels:      rayService.Labels,
+			Annotations: rayService.Annotations,
+			Name:        name,
+			Namespace:   rayService.Namespace,
+		},
+		Spec: *rayService.Spec.RayClusterSpec.DeepCopy(),
+	}
+
+	// Set the ownership in order to do the garbage collection by k8s.
+	if err := ctrl.SetControllerReference(rayService, rayCluster, r.Scheme); err != nil {
+		return nil, err
+	}
+
+	return rayCluster, nil
+}
+
+func (r *RayServiceReconciler) fetchDashboardURL(rayCluster *rayv1alpha1.RayCluster) (string, error) {
+	headServices := corev1.ServiceList{}
+	filterLabels := client.MatchingLabels{common.RayClusterLabelKey: rayCluster.Name}
+	if err := r.List(context.TODO(), &headServices, client.InNamespace(rayCluster.Namespace), filterLabels); err != nil {
+		return "", err
+	}
+
+	dashboardURL := ""
+
+	if headServices.Items != nil {
+		if len(headServices.Items) == 1 {
+			r.Log.Info("reconcileServices ", "head service found", headServices.Items[0].Name)
+			// TODO: compare diff and reconcile the object. For example. ServiceType might be changed or port might be modified
+			servicePorts := headServices.Items[0].Spec.Ports
+
+			dashboardPort := int32(-1)
+
+			for _, servicePort := range servicePorts {
+				if servicePort.Name == "dashboard" {
+					dashboardPort = servicePort.Port
+					break
+				}
+			}
+
+			if dashboardPort == int32(-1) {
+				return "", fmtErrors.Errorf("dashboard port not found")
+			}
+
+			dashboardURL = fmt.Sprintf("%s.%s.svc.cluster.local:%v",
+				headServices.Items[0].Name,
+				headServices.Items[0].Namespace,
+				dashboardPort)
+			return dashboardURL, nil
+		}
+
+		// This should never happen.
+		// We add the protection here just in case controller has race issue or user manually create service with same label.
+		if len(headServices.Items) > 1 {
+			r.Log.Info("reconcileServices ", "Duplicates head service found", len(headServices.Items))
+			return "", fmtErrors.Errorf("Duplicates head service found %v", len(headServices.Items))
+		}
+	}
+
+	// Create head service if there's no existing one in the cluster.
+	if headServices.Items == nil || len(headServices.Items) == 0 {
+		return "", fmtErrors.Errorf("No head service found")
+	}
+
+	return "", nil
+}
+
+func (r *RayServiceReconciler) checkIfNeedSubmitServeDeployment(rayServiceInstance *rayv1alpha1.RayService, request ctrl.Request) bool {
+	existConfig, exist := r.ServeDeploymentConfigMap[request.NamespacedName]
+
+	shouldUpdate := false
+
+	if !exist || !reflect.DeepEqual(existConfig, rayServiceInstance.Spec) || len(rayServiceInstance.Status.ServeStatuses.Statuses) != len(existConfig.ServeConfigSpecs) {
+		shouldUpdate = true
+	}
+	r.Log.Info("shouldUpdate value", "shouldUpdate", shouldUpdate)
+
+	log.V(1).Info("status check", "len(rayServiceInstance.Status.ServeStatuses.Statuses) ", len(rayServiceInstance.Status.ServeStatuses.Statuses), "len(existConfig.ServeConfigSpecs)", len(existConfig.ServeConfigSpecs))
```

Comment:
For better log readability, maybe add a colon after "len(rayServiceInstance.Status.ServeStatuses.Statuses)" and a comma before the second len.

Reply:
I suppose to delete these logs in this pr.